File: codecheck-case/lua/BKCT_ForwardNull/BKCT_ForwardNull_01.lua
local BKCT_ForwardNull_01 = {}

function BKCT_ForwardNull_01:badMethod1()
    local str = nil
    if str == nil then
        -- bad
        print(#str) -- 在为空的分支中使用
    end
end

function BKCT_ForwardNull_01:badMethod2()
    local str = nil
    if str ~= nil then
        return
    end
    -- bad
    print(#str) -- 在为空的分支中使用
end

function BKCT_ForwardNull_01:badMethod3(str)
    if str ~= nil then
        str = nil
    end
    -- bad
    print(#str) -- 在不再次检查是否为空的情况下使用
end

function BKCT_ForwardNull_01:badMethod4(str)
    if str == nil then
        -- do nothing
    end
    -- bad
    print(#str) -- 在不再次检查是否为空的情况下使用
end

function BKCT_ForwardNull_01:goodMethod1()
    local str = nil
    if str ~= nil then
        -- good
        print(#str)
    end
end

function BKCT_ForwardNull_01:goodMethod2()
    local str = nil
    if str == nil then
        return
    end
    -- good
    print(#str)
end

function BKCT_ForwardNull_01:goodMethod3(str)
    if str == nil then
        str = "aaa"
    end
    -- good
    print(#str)
end

function BKCT_ForwardNull_01:goodMethod4(str)
    if str == nil then
        return
    end
    -- good
    print(#str)
end

return BKCT_ForwardNull_01

File: codecheck-case/cs/BKCT_ForwardNull/BKCT_ForwardNull_01.cs
using System;

public class BKCT_ForwardNull_01
{
    public void BadMethod1()
    {
        string str = null;
        if (str == null)
        {
            // bad
            Console.WriteLine(str.Length); // 在为空的分支中使用
        }
    }

    public void BadMethod2()
    {
        string str = null;
        if (str != null)
        {
            return;
        }
        // bad
        Console.WriteLine(str.Length); // 在为空的分支中使用
    }

    public void BadMethod3(string str)
    {
        if (str != null)
        {
            str = null;
        }
        // bad
        Console.WriteLine(str.Length); // 在不再次检查是否为空的情况下使用
    }

    public void BadMethod4(string str)
    {
        if (str == null)
        {
            // do nothing
        }
        // bad
        Console.WriteLine(str.Length); // 在不再次检查是否为空的情况下使用
    }

    public void GoodMethod1()
    {
        string str = null;
        if (str != null)
        {
            // good
            Console.WriteLine(str.Length);
        }
    }

    public void GoodMethod2()
    {
        string str = null;
        if (str == null)
        {
            return;
        }
        // good
        Console.WriteLine(str.Length);
    }

    public void GoodMethod3(string str)
    {
        if (str == null)
        {
            str = "aaa";
        }
        // good
        Console.WriteLine(str.Length);
    }

    public void GoodMethod4(string str)
    {
        if (str == null)
        {
            return;
        }
        // good
        Console.WriteLine(str.Length);
    }
}